File: backend/src/main/java/com/sentinela/alpr/watchlist/domain/WatchlistService.java
package com.sentinela.alpr.watchlist.domain;

import java.util.Optional;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.sentinela.alpr.shared.error.BusinessRuleException;
import com.sentinela.alpr.shared.error.ConflictException;
import com.sentinela.alpr.shared.error.NotFoundException;
import com.sentinela.alpr.shared.support.PlateNormalizer;
import com.sentinela.alpr.watchlist.api.WatchlistRequest;
import com.sentinela.alpr.watchlist.api.WatchlistResponse;
import com.sentinela.alpr.watchlist.api.WatchlistSummaryResponse;
import com.sentinela.alpr.watchlist.api.WatchlistUpdateRequest;
import com.sentinela.alpr.watchlist.infra.WatchedVehicleRepository;

@Service
public class WatchlistService {

	private final WatchedVehicleRepository repository;

	WatchlistService(WatchedVehicleRepository repository) {
		this.repository = repository;
	}

	@Transactional
	public WatchlistResponse add(WatchlistRequest request) {
		String plate = normalizePlate(request.plate());
		if (repository.existsByPlate(plate)) {
			throw new ConflictException("Plate already in watchlist: " + plate);
		}
		boolean active = request.active() == null || request.active();
		WatchedVehicle saved = repository.save(new WatchedVehicle(plate, request.reason(), active));
		return toResponse(saved);
	}

	@Transactional(readOnly = true)
	public Optional<WatchlistResponse> findById(Long id) {
		return repository.findById(id).map(WatchlistService::toResponse);
	}

	@Transactional(readOnly = true)
	public Page<WatchlistResponse> list(Pageable pageable) {
		return repository.findAll(pageable).map(WatchlistService::toResponse);
	}

	@Transactional(readOnly = true)
	public WatchlistSummaryResponse summary() {
		return new WatchlistSummaryResponse(repository.countByActiveTrue());
	}

	@Transactional
	public WatchlistResponse update(Long id, WatchlistUpdateRequest request) {
		WatchedVehicle vehicle = require(id);
		vehicle.setReason(request.reason());
		return toResponse(vehicle);
	}

	@Transactional
	public void deactivate(Long id) {
		require(id).setActive(false);
	}

	@Transactional
	public void activate(Long id) {
		require(id).setActive(true);
	}

	@Transactional(readOnly = true)
	public Optional<WatchlistResponse> findActiveMatch(String plate) {
		return repository.findByPlateAndActiveTrue(PlateNormalizer.normalize(plate))
				.map(WatchlistService::toResponse);
	}

	private WatchedVehicle require(Long id) {
		return repository.findById(id)
				.orElseThrow(() -> new NotFoundException("Watchlist item not found: " + id));
	}

	private static String normalizePlate(String raw) {
		String plate = PlateNormalizer.normalize(raw);
		if (!PlateNormalizer.isValid(plate)) {
			throw new BusinessRuleException("Plate is invalid: " + raw);
		}
		return plate;
	}

	private static WatchlistResponse toResponse(WatchedVehicle w) {
		return new WatchlistResponse(w.getId(), w.getPlate(), w.getReason(), w.isActive(), w.getCreatedAt());
	}
}

File: frontend/src/app/core/watchlist/watchlist.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';

import { Page } from '../../shared/models/page.model';
import {
  WatchedVehicle,
  WatchlistRequest,
  WatchlistUpdateRequest,
} from '../../shared/models/watchlist.model';

@Injectable({ providedIn: 'root' })
export class WatchlistService {
  private readonly http = inject(HttpClient);
  private readonly base = '/api/v1/watchlist';

  list(page = 0, size = 20): Observable<Page<WatchedVehicle>> {
    const params = new HttpParams().set('page', page).set('size', size);
    return this.http.get<Page<WatchedVehicle>>(this.base, { params });
  }

  create(request: WatchlistRequest): Observable<WatchedVehicle> {
    return this.http.post<WatchedVehicle>(this.base, request);
  }

  update(id: number, request: WatchlistUpdateRequest): Observable<WatchedVehicle> {
    return this.http.put<WatchedVehicle>(`${this.base}/${id}`, request);
  }

  deactivate(id: number): Observable<void> {
    return this.http.delete<void>(`${this.base}/${id}`);
  }

  activate(id: number): Observable<void> {
    return this.http.post<void>(`${this.base}/${id}/activate`, {});
  }
}
